cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/StopStartUpscaleStartInstancesHandler.java

```diff
@@ -42,59 +47,131 @@ public void accept(Event<StopStartUpscaleStartInstancesRequest> event) {
+            notify(result, event);
+            return;
+        }
+
         CloudContext cloudContext = request.getCloudContext();
         try {
             CloudConnector<?> connector = cloudPlatformConnectors.get(cloudContext.getPlatformVariant());
             AuthenticatedContext ac = getAuthenticatedContext(request, cloudContext, connector);
 
-            List<CloudInstance> stoppedInstancesInHg = request.getStoppedCloudInstancesInHg();
-            int numInstancesToStart = request.getNumInstancesToStart() < stoppedInstancesInHg.size() ?
-                    request.getNumInstancesToStart() : stoppedInstancesInHg.size();
-            LOGGER.info("ZZZ: Requested instances to start: {}, actual instances being started: {}", request.getNumInstancesToStart(), numInstancesToStart);
+            List<CloudInstance> stoppedInstancesInCbHg = request.getStoppedCloudInstancesInHg();
+            List<CloudVmInstanceStatus> stoppedInstancesOnCloudProvider = Collections.emptyList();
 
-            // TODO CB-14929: This should ideally be randomized a bit, so that different isntances are started/stopped each time. That said, there is value in
-            //  a reliable pattern. i.e. something along the lines of sort by hostname within this list.
+            List<CloudInstance> instancesToStart = Collections.emptyList();
 
-            List<CloudInstance> instancesToStart = request.getStoppedCloudInstancesInHg().subList(0, numInstancesToStart);
-            // TODO CB-14929: Additional validation here to check if these instances are actually in STOPPED state
-            //  (as against the current list which is an indication of what the Cloudbreak Database thinks are STOPPED instances)
-
-            LOGGER.info("ZZZ: Instance identified as start candidates: count={}, instances={}", instancesToStart.size(), instancesToStart);
-            if (instancesToStart.size() < request.getNumInstancesToStart()) {
-                LOGGER.info("ZZZ: There are fewer instances available to start as compared to the request. Requested: {}, Available: {}",
-                        request.getNumInstancesToStart(), instancesToStart.size());
+            if (numInstancesToStart < stoppedInstancesInCbHg.size()) {
+                // See if we can find additional instances on the cloud-provider which can potentially be STARTED.
+                stoppedInstancesOnCloudProvider = collectStoppedInstancesFromCloudProvider(connector, ac, request.getAllInstancesInHg());
+                instancesToStart = getInstancesToStart(stoppedInstancesInCbHg, stoppedInstancesOnCloudProvider, request.getHostGroupName(), numInstancesToStart);
+            } else {
+                // TODO CB-15132: Potentially validate against the cloud-provider to make sure these instances are actually in STOPPDE state.
+                // Try starting the required instances based on CB state.
+                instancesToStart = stoppedInstancesInCbHg.subList(0, numInstancesToStart);
```

Comment:
Info: This is good because the node count validator will fail if there are more instances required to start than what is currently available in the stopped state.